# Project

# 1. About Dataset 

## 1-1) 데이터 선정사유 
  

최근 기후 변화로 와인의 생산량이 급감하고 있는 문제가 발생되고 있다.  
이러한 현상으로 와인 풍귀현상 및 사재기 현상도 함께 발생되어 사회문제로 부각되고 있다.  

이에, 미국, 프랑스, 이테리, 대한민국 정부에서는 와인 판매를 국가에서 통제하고자 한다.  
정부 담당자는 판매되는 와인의 가격으로 결정하기 위해 기존에 판매된 와인가격에 대한 데이터 분석을  
요청하였다.  


DS.ltd 와인 가격분석을 진행하고자 합니다.  


---

## 1-2) target 선정사유 

  
우리는 기존에 판매된 와인가격에 대한 분석을 진행해야 하므로 price를 target으로 선정토록 하겠습니다.  
  
또한, price를 기준으로 어떤 와인이 좋다 나쁘다의 분류를 하는 것이 아니며   
적절한 가격은 얼마정도인지 가격 예측을 목표로 하기 때문에 회귀분석으로 접근토록 하겠습니다. 

---

## 1-3) 데이터 로드 및 확인  

In [1]:
import pandas as pd
import numpy as np 

In [2]:
#df2 = pd.read_csv(r"C:\Users\user\Documents\pandas_study\bootcamp_DS\project\section2_porject_210222\winemag-data_first150k.csv",  index_col=0)

In [3]:
df1 = pd.read_csv(r"C:\Users\user\Documents\pandas_study\bootcamp_DS\project\section2_porject_210222\winemag-data-130k-v2.csv", index_col=0)

In [4]:
df1.head(50)

country                                        description  \
0       Italy  Aromas include tropical fruit, broom, brimston...   
1    Portugal  This is ripe and fruity, a wine that is smooth...   
2          US  Tart and snappy, the flavors of lime flesh and...   
3          US  Pineapple rind, lemon pith and orange blossom ...   
4          US  Much like the regular bottling from 2012, this...   
5       Spain  Blackberry and raspberry aromas show a typical...   
6       Italy  Here's a bright, informal red that opens with ...   
7      France  This dry and restrained wine offers spice in p...   
8     Germany  Savory dried thyme notes accent sunnier flavor...   
9      France  This has great depth of flavor with its fresh ...   
10         US  Soft, supple plum envelopes an oaky structure ...   
11     France  This is a dry wine, very spicy, with a tight, ...   
12         US  Slightly reduced, this wine offers a chalky, t...   
13      Italy  This is dominated by oak and oak-driven aromas...   
14         US  Building on 150 years and six generations of w...   
15    Germany  Zesty orange peels and apple notes abound in t...   
16  Argentina  Baked plum, molasses, balsamic vinegar and che...   
17  Argentina  Raw black-cherry aromas are direct and simple ...   
18      Spain  Desiccated blackberry, leather, charred wood a...   
19         US  Red fruit aromas pervade on the nose, with cig...   
20         US  Ripe aromas of dark berries mingle with ample ...   
21         US  A sleek mix of tart berry, stem and herb, alon...   
22      Italy  Delicate aromas recall white flower and citrus...   
23         US  This wine from the Geneseo district offers aro...   
24      Italy  Aromas of prune, blackcurrant, toast and oak c...   
25         US  Oak and earth intermingle around robust aromas...   
26      Italy  Pretty aromas of yellow flower and stone fruit...   
27      Italy  Aromas recall ripe dark berry, toast and a whi...   
28      Italy  Aromas suggest mature berry, scorched earth, a...   
29         US  Clarksburg is becoming a haven for Chenin Blan...   
30     France  Red cherry fruit comes laced with light tannin...   
31      Italy  Merlot and Nero d'Avola form the base for this...   
32      Italy  Part of the extended Calanìca series, this Gri...   
33         US  Rustic and dry, this has flavors of berries, c...   
34         US  This shows a tart, green gooseberry flavor tha...   
35         US  As with many of the Erath 2010 vineyard design...   
36      Chile  White flower, lychee and apple aromas carry th...   
37      Italy  This concentrated Cabernet offers aromas of cu...   
38      Italy  Inky in color, this wine has plump aromas of r...   
39      Italy  Part of the natural wine movement, this wine i...   
40      Italy  Catarratto is one of Sicily's most widely farm...   
41         US  A stiff, tannic wine, this slowly opens and br...   
42     France  This is a festive wine, with soft, ripe fruit ...   
43         US  The clean, brisk mouthfeel gives this slightly...   
44      Chile  A berry aroma comes with cola and herb notes. ...   
45         US  Right out of the starting blocks this is an oa...   
46      Italy  Spicy, fresh and clean, this would pair with f...   
47         US  This is a sweet wine with flavors of white sug...   
48         US  This bottling resembles the New Zealand paradi...   
49     France  Soft and fruity, this is a generous, ripe wine...   

                                          designation  points  price  \
0                                        Vulkà Bianco      87    NaN   
1                                            Avidagos      87   15.0   
2                                                 NaN      87   14.0   
3                                Reserve Late Harvest      87   13.0   
4                  Vintner's Reserve Wild Child Block      87   65.0   
5                                        Ars In Vitro      87   15.0   
6                                

In [5]:
# 데이터량 확인
df1.shape

(129971, 13)

In [6]:
# 데이터 결측치 확인 
df1.isnull().sum()

country                     63
description                  0
designation              37465
points                       0
price                     8996
province                    63
region_1                 21247
region_2                 79460
taster_name              26244
taster_twitter_handle    31213
title                        0
variety                      1
winery                       0
dtype: int64

In [7]:
# df1.info로 확인 결과 대부분 object 데이터 타입임을 확인하였다. 
# check the data info -> This data is almost object types 
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 129971 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            92506 non-null   object 
 3   points                 129971 non-null  int64  
 4   price                  120975 non-null  float64
 5   province               129908 non-null  object 
 6   region_1               108724 non-null  object 
 7   region_2               50511 non-null   object 
 8   taster_name            103727 non-null  object 
 9   taster_twitter_handle  98758 non-null   object 
 10  title                  129971 non-null  object 
 11  variety                129970 non-null  object 
 12  winery                 129971 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 13.9+ MB


In [8]:
# object data의 unique를 살펴보았다. 
df1.describe(exclude='number').T.sort_values(by='unique')

count  unique  \
taster_twitter_handle   98758      15   
region_2                50511      17   
taster_name            103727      19   
country                129908      43   
province               129908     425   
variety                129970     707   
region_1               108724    1229   
winery                 129971   16757   
designation             92506   37979   
title                  129971  118840   
description            129971  119955   

                                                                     top  \
taster_twitter_handle                                         @vossroger   
region_2                                                   Central Coast   
taster_name                                                   Roger Voss   
country                                                               US   
province                                                      California   
variety                                                       Pinot Noir   
region_1                                                     Napa Valley   
winery                                                Wines & Winemakers   
designation                                                      Reserve   
title                  Gloria Ferrer NV Sonoma Brut Sparkling (Sonoma...   
description            Cigar box, café au lait, and dried tobacco aro...   

                        freq  
taster_twitter_handle  25514  
region_2               11065  
taster_name            25514  
country                54504  
province               36247  
variety                13272  
region_1                4480  
winery                   222  
designation             2009  
title                     11  
description                3

## 1-4) feature 설명 
  
'country' - 어느나라에서 생산되었는지  
'description' - 와인 시음 후 와인에 대한 평가 및 설명   
'designation' - 양조장이 있는 포도원   
'points' - 시음 후 와인 점수   
'price' - 와인의 가격 ( "our target" )  
'province' - 와인 제작 시, 사용되는 포도의 출처   
'region_1' - 포도가 자라는 지역 
'region_2' - 포도가 자라는 지역을 상세하게 기재한 feature
'taster_name' - 시음한 사람     
'taster_twitter_handle' 시음한 사람의 트위터   
'title' - 와인명  
'variety' - 와인을 만들때 혼합된 포도  
'winery' - 양조장명  

### **☆ 지금까지 insight ☆**  
 1. designation과 winery를 동일하게 양조장에 대한 정보를 나타내고 있다.  
     -> 확인 후 합칠 수 있는 것들은 혼합할 예정   
       
 2. region_1과 region_2도 동일한 정보를 나타내며, title에 region정보가 있는 것을 확인할 수 있었다.  
    -> 결측치는 이로 대체 예정 
      
 3. taster_name와 taster_twitter_handle도 taster에 대한 동일한 정보를 제공해주고 있다.  
    -> 확인 후 합칠 수 있는 것들은 혼합할 예정 
      
 4. title, variety, winery 등은 결측치가 거의 없다. 이에, 해당 데이터를 활용하여 결측치를 채울 수 있는 것들은 채울 예정
   
 5. description을 split하여 aroma, blend, sweet 등의 단어를 찾아서 타겟과의 연관관계를 확인해보고자 한다    
   
 6. 또한, 너무 unique가 높기 때문에 각 feature별로 count하여 count된 만큼 페센트로 가중치를 주고자 한다.  


### ☆ 나라별 용어 따른 숙성 정도 차이 ☆

7. 스페인, 이탈리아 ( 법적 필수 준수사항 )


 |등급|내용|
 |----|-----|
 |joven|전혀 숙성을 시키지 않고 바로 출시된 와인|
 |crianza|최소 24개월이상 오크 및 병에 담아 숙성 후 출시된 와인|
 |reserva|최소 36개월이상 오크 및 병에 담아 숙성 후 출시된 와인|
 |gran reserva|최소 60개월이상 오크 및 병에 담아 숙성 후 출시된 와인|

8. 그외나라(미국, 칠레, 호주, 뉴질랜드 등)   
 -. 별다른 규정 없이 바로 taster에게 좋은 평가를 받으면 reserva라는 표기 가능   

9. 기타(혼용되는 용어)  
 -. reserva는 나라에 따라서 riserva, reserve 등으로 혼용되어 사용된다고 한다.  
 -. ["À", "A"],["Á", "A"],["Â", "A"],["Ã", "A"],["Ä", "A"],["Ä", "A"],["Å", "A"]  
 -. ,["à", "a"],["á", "a"],["â", "a"],["ã", "a"],["ä", "a"],["ä", "a"],["å", "a"]  
 -. ,["È", "E"],["É", "E"],["Ê", "E"],["Ë", "E"],["Ë", "E"]  
 -. ,["è", "e"],["é", "e"],["ê", "e"],["ë", "e"],["ë", "e"]  
 -. ,["Æ", "AE"],["æ", "ae"],["Ç", "C"],["ç", "c"]  
 -. ,["Ì", "I"],["Í", "I"],["Î", "I"],["Ï", "I"],["Ï", "I"]  
 -. ,["ì", "i"],["í", "i"],["î", "i"],["ï", "i"],["ï", "i"]  
 -. ,["Ð", "D"],["ð", "d"],["Ñ", "N"],["ñ", "n"]  
 -. ,["Ò", "O"],["Ó", "O"],["Ô", "O"],["Õ", "O"],["Ö", "O"],["Ö", "O"],["Œ", "OE"]   
 -. ,["ò", "o"],["ó", "o"],["ô", "o"],["õ", "o"],["ö", "o"],["ö", "o"],["œ", "oe"]  
 -. ,["Ù", "U"],["Ú", "U"],["Û", "U"],["Ü", "U"],["Ü", "U"]  
 -. ,["ù", "u"],["ú", "u"],["û", "u"],["ü", "u"],["ü", "u"]  
 -. ,["Ý", "Y"],["Ÿ", "Y"],["Ÿ", "Y"]  
 -. ,["ý", "y"],["ÿ", "y"],["ÿ", "y"]  
 -. ,["Þ", "P"],["þ", "p"],["ß", "s"]["¡", "!"]["¿", "?"]  
  -> 위 용어를 모두 분석한 결과, 부분적으로만 대문자로 시작되는 글자는 없었으며,    
  -> 제품의 발음을 위해 France 등 일부 유럽국가에서만 특수문자로 기재가 되어 있음을 확인하였다.   
  -> 이에, 카테고리티를 줄이기 위해서 이를 통일하는 작업이 필요하다가 생각된다. 

10. 출처   
 -. 와인등급 :https://www.wine21.com/11_news/news_view.html?Idx=17111  
 -. 와인등급 : https://m.blog.naver.com/interkj87/220699483883  
 -. 와인등급 : https://www.decantalo.com/en/blog/winemaking/classifying-wines-by-their-age-crianza-reserva-gran-reserva/  
 -. 유럽(문자) 표기법:http://help.hancom.com/hoffice/multi/ko_kr/hwp/insert/europe.htm
※ 모든 title에 해당 정보가 기재되어 있음

유럽(외국문자) 표기법
http://help.hancom.com/hoffice/multi/ko_kr/hwp/insert/europe.htm

In [9]:
# 혼용되는 용어 통일을 위한 object 추출 
df1.select_dtypes(exclude="number").columns

Index(['country', 'description', 'designation', 'province', 'region_1',
       'region_2', 'taster_name', 'taster_twitter_handle', 'title', 'variety',
       'winery'],
      dtype='object')

In [10]:
# 혼용되는 용어 통일을 위한 object 리스트화
object_columns = ['country', 'description', 'designation', 'province', 'region_1',
       'region_2', 'taster_name', 'taster_twitter_handle', 'title', 'variety',
       'winery']

In [11]:
# 문자 영어로 통일 
lists = [["À","Á","Â","Ã","Ä","Ä","Å"]
        ,["à","á","â","ã","ä","ä","å"]
        ,["È","É","Ê","Ë","Ë"]
        ,["è", "é", "ê", "ë", "ë"]     
        ,["ç", "c"]
        ,["ì","í","î","ï","ï"]
        ,["ò","ó","ô","õ","ö","ö","œ"]
        ,["ñ"]
        ,["ù","ú","û","ü","ü"]
        ,["ý","ÿ","ÿ"]]

for list in lists[0][1] :
    df1 = df1.replace(list, "a", regex=True)

for list in lists[2][3] :
    df1 = df1.replace(list, "e", regex=True)
    
for list in lists[4] :
    df1 = df1.replace(list, "c", regex=True)
    
for list in lists[5] :
    df1 = df1.replace(list, "i", regex=True)
    
for list in lists[6] :
    df1 = df1.replace(list, "o", regex=True)
    
for list in lists[7] :
     df1 = df1.replace(list, "n", regex=True)
    
for list in lists[8] :
    df1 = df1.replace(list, "u", regex=True)
    
for list in lists[9] :
    df1 = df1.replace(list, "y", regex=True)

In [12]:
df1["designation"].isnull().sum()

37465

In [13]:


df1 = df1.dropna(axis=0)


In [14]:
df1["designation"].isnull().sum()

37465

In [ ]:
# 모든 대문자 -> 소문자
for column in object_columns:
    df1[column] = df1[column].str.lower()

In [15]:
df1.shape

(129971, 13)

In [16]:
df1

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
0                                 Vulkà Bianco      87    NaN   
1                                     Avidagos      87   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2  \
0       Sicily & Sardinia                 Etna                NaN   
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle  \
0            Kerin O’Keefe          @kerinokeefe   
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Muller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewurztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewurztraminer 

In [17]:
# 37979 -> 37609로 unique 감소 
df1["designation"].describe()

count       92506
unique      37869
top       Reserve
freq         2009
Name: designation, dtype: object

In [18]:
for column in object_columns:
    df1[column] = df1[column].str.replace("reserva","reserve")

for column in object_columns:
    df1[column] = df1[column].str.replace("riserva","reserve")

In [19]:
# 37609에서 37562로 unique 감소 
df1["designation"].describe()

count       92506
unique      37869
top       Reserve
freq         2009
Name: designation, dtype: object

아래와 같이 "designation","title" 살펴본 결과, title의 연도부터 괄호 사이에 있는 내용이 designation임을 확인 할 수 있었다. 이로 designation의 결측치 37,465개를 처리할 수 있다.

In [20]:
joven=df1[["designation","title"]].query('designation.str.contains("joven",  na=False)',engine='python')
crianza=df1[["designation","title"]].query('designation.str.contains("crianza",  na=False)',engine='python')
reserve=df1[["designation","title"]].query('designation.str.contains("reserve",  na=False)',engine='python')
gran_reserve=df1[["designation","title"]].query('designation.str.contains("gran reserve",  na=False)',engine='python')

In [21]:
len(joven)+len(crianza)+len(reserve)+len(gran_reserve)

1

In [22]:
len(df1["designation"])

129971

In [24]:
for grade in df1["designation"]:
    if "joven" in grade:
        df1["designation"] = df1["designation"].str.replace(grade, "joven")

In [25]:
for grade in df1["designation"]:
    if "crianza" in grade:
        df1["designation"] = df1["designation"].str.replace(grade, "crianza")

In [26]:
for grade in df1["designation"]:
    if "gran reserve" in grade:
        df1["designation"] = df1["designation"].str.replace(grade, "gran reserve")

In [27]:
for grade in df1["designation"]:
    if "reserve" in grade:
        df1["designation"] = df1["designation"].str.replace(grade, "reserve")

In [29]:
gran_reserve=df1[["designation","title"]].query('designation.str.contains("nan",  na=False)',engine='python')

In [35]:
df1[["region_1","title"]].query('title.str.contains("joven")',engine='python').head(10)

Empty DataFrame
Columns: [region_1, title]
Index: []

위와 같이 데이터를 살펴본 결과, title의 괄호 부분이 region_1를 나타냄을 알 수 있었다.
이로 region_1의 결측치 21,247개를 처리할 수 있다.

In [38]:
df1["title"] = df1["title"].str.split("(")
df1["title"] = df1["title"].str.split(")")

In [39]:
df1["title"]

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
          ..
129966   NaN
129967   NaN
129968   NaN
129969   NaN
129970   NaN
Name: title, Length: 129971, dtype: float64

In [33]:
df1['price'].describe()

count    120975.000000
mean         35.363389
std          41.022218
min           4.000000
25%          17.000000
50%          25.000000
75%          42.000000
max        3300.000000
Name: price, dtype: float64

In [34]:
df1[["title"]].query('title.str.contains("riserva")',engine='python').head(10)

Empty DataFrame
Columns: [title]
Index: []